### PR TITLE
resolve ambiguity between colon methods

### PR DESCRIPTION
**Pull request by Sacha0**
Someone familiar with the range code should have a look; chances are a more graceful fix exists. Best!

### CODE REVIEW

**Comment by Sacha0** [comment]
(AV failure is the familiar `pkg` test issue.)

**Comment by timholy** [comment]
For 0.6 I suspect this is the correct solution (don't want to change the dispatch hierarchy at this stage). For 1.0, I wonder if we could get away with less typing on the `promote`ing methods, e.g.,
```julia
colon(a, b, c) = (ap, cp = promote_noncircular(a, c); colon(ap, b, cp))
colon{T}(a::T, b, c::T) = (ap, bp, cp = promote_noncircular(a, b, c); colon(ap, bp, cp))
```
`colon` implicitly requires ordering, which `Real` provides, but other types can also be ordered. So perhaps we should only be imposing constraints once we get to specific methods.

File: test/ranges.jl [hunk 1]
```diff
@@ -894,6 +894,14 @@ let r = linspace(1.0, 3+im, 4)
     @test r[4] === 3.0+im
 end
 
+# ambiguity between colon methods (#20988)
+struct NotReal; val; end
+Base.:+(x, y::NotReal) = x + y.val
+Base.zero(y::NotReal) = zero(y.val)
+Base.rem(x, y::NotReal) = rem(x, y.val)
+Base.isless(x, y::NotReal) = Base.isless(x, y.val)
```
Comment (tkelman) [inline test/ranges.jl]: is `Base.` needed on the right hand side?

Comment (Sacha0) [inline test/ranges.jl]: Seems not :). Thanks!

**Comment by Sacha0** [comment]
Thanks!

**Pull request merged**